### ISSUE TITLE
Handle batch queries in `DbApiHook`?

### ISSUE THREAD

**Issue by eladkal** [apache/airflow#23112]
### Body

**TL;DR:**
Multiple providers need to handle the logic of splitting SQL statements
[Snowflake.run()](https://github.com/apache/airflow/blob/501a3c3fbefbcc0d6071a00eb101110fc4733e08/airflow/providers/snowflake/hooks/snowflake.py#L287-L347) , [RedShift.run()](https://github.com/apache/airflow/pull/22391/files) and it will be very similar should be eventually create `PrestoOperator` or `TrinoOperator`.

I want to discuss the idea of extracting the split statement logic into `DbApiHook`.
e.g create a `DbApiHook.run_many()` that will split the statements and use `DbApiHook.run()` under the hood.

On one hand it will reduce duplicate code and allow to develop operators for these providers more easily
On the other hand if we will make it in `dbapihook` it means to bump minimum supported Airflow version for these providers (once they will actually use the new function)

----------------------------
**The long story:**
Some DBs like MySQL, PostgreSql, MsSQL are able to run a batch of SQL statements:
`SELECT 1; SELECT2;`

So you can do:
`MySQLOperator(.., sql='SELECT 1; SELECT2;')`
it will invoke :
https://github.com/apache/airflow/blob/501a3c3fbefbcc0d6071a00eb101110fc4733e08/airflow/hooks/dbapi.py#L163

which will execute the sql in a single query. (The run function will convert this sql to `['SELECT 1; SELECT2;']`  - Note this is a list with **1 cell**)

But in other DBs like: Snowflake, Redshift, Trino, Presto

You can not run `SELECT 1; SELECT2;` as is. You must split the statement.
This makes the `run()` irrelevant for these DBs.

For example in Snowflake when users pass `SELECT 1; SELECT2;` what we actually do is create:
`['SELECT 1;', 'SELECT2;']` - Note this is a list with **2 cells**
https://github.com/apache/airflow/blob/501a3c3fbefbcc0d6071a00eb101110fc4733e08/airflow/providers/snowflake/hooks/snowflake.py#L314-L316

**The problem?**
These providers (Snowflake, Redshift, Trino, Presto) **are forced to override dbapi.run()** but they all actually need the exact same functionality.

If you will take a look at [Snowflake.run()](https://github.com/apache/airflow/blob/501a3c3fbefbcc0d6071a00eb101110fc4733e08/airflow/providers/snowflake/hooks/snowflake.py#L287-L347) you will see that it's almost identical to [RedShift.run()](https://github.com/apache/airflow/pull/22391/files) and it will be very similar should be eventually create `PrestoOperator` or `TrinoOperator`.

This is something already brought up earlier in https://github.com/apache/airflow/pull/15533#discussion_r620481685

**So why not just set the sql param to accept list only?**
Because we want to support using .sql file so we must handle the split of the statements.
```
SomeSqlOperator(
...,
sql=my_queries.sql
)
```
We are reading the content of the file and loading it as a single statement.

--------------------------

**What do others think?**
Should we create `DbApiHook.run_many()` or leave it as it is today where every provider needs to handle the logic on it's own?

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

**Comment by potiuk** [comment]
It would be good to standardize it indeed. And temporary implementation that would use the function (if existing in Aifrlow) and falling back to internal, deprecated implementation is the way to go. We can even add pre-commit guarding the proper use of (run_many) - making sure that there is a conditional code in provider that falls back if run_many is used (new operators will use it automatically when present I believe).

**Comment by josh-fell** [comment]
+1 and completely agree with @potiuk. An underlying standard would be very useful and less repetitive.

**Comment by williesb** [comment]
What is the tentative ETA on this?  I am stuck at the moment not being able to run multiple statements in a single connection/session with the TrinoHook because the run implementation does not take a list like DBApiHook does.

Can we please make sure that this new method uses a single connection/session to run all the statements like the current run methods does with statement lists?

**Comment by eladkal** [comment]
@williesb currently no one is working on this. Would you like to start a PR?

**Comment by williesb** [comment]
> @williesb currently no one is working on this. Would you like to start a PR?

@eladkal It sounds like a tempting idea.  Your original post makes sense.  However, thinking more about it, if it was implemented as you suggest that the `DbApiHook.run_many()` would split the statements and use `DbApiHook.run()` under the hood.  That probably wouldn't fix my problem (I am not 100% not a Python expert) because the `TrinoHook.run()` methods does not work with lists.  I wish @uranusjr would have closed the issue I opened so quickly https://github.com/apache/airflow/issues/23431.  That I can fix quickly.  This one might have some nuances that are a bit complicated for me to resolve on my own.

**Comment by kevgeo** [comment]
Hi @eladkal, could I take up this issue?

**Comment by eladkal** [comment]
We are waiting for https://github.com/apache/airflow/pull/23971 to be finished first to see what actions needs to be taken here

**Comment by eladkal** [comment]
closed by https://github.com/apache/airflow/pull/23971 (added the `split_statements` parameter to `run()` )
